bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -24,6 +24,158 @@
 using namespace std::chrono;
 using namespace concord::serialize;
 namespace bftEngine::impl {
+
+/*************************** Class ClientsManager::RequestsInfo ***************************/
+
+void ClientsManager::RequestsInfo::emplaceSafe(NodeIdType clientId, ReqId reqSeqNum, const std::string& cid) {
+  const lock_guard<mutex> lock(requestsMapMutex_);
+  if (requestsMap_.find(reqSeqNum) != requestsMap_.end()) {
+    LOG_WARN(CL_MNGR, "The request already exists - skip adding" << KVLOG(clientId, reqSeqNum));
+    return;
+  }
+  requestsMap_.emplace(reqSeqNum, RequestInfo{getMonotonicTime(), cid});
+  LOG_DEBUG(CL_MNGR, "Added request" << KVLOG(clientId, reqSeqNum, requestsMap_.size()));
+}
+
+bool ClientsManager::RequestsInfo::findSafe(ReqId reqSeqNum) {
+  const lock_guard<mutex> lock(requestsMapMutex_);
```

Comment:
Taking the write-only mutex is a good idea. But here we can see that we have separate read and write operations on requestMap_. So a read and write lock will be better for better performance:
If we use the read and write lock as below:
using WriteLock = std::unique_lock<std::shared_mutex>;
using ReadLock = std::shared_lock<std::shared_mutex>;

Then it will keep the throughput better.
So in emplaceSafe() we should use WriteLock(requestsMapMutex_);
And in findSafe() we should use ReadLock(requestsMapMutex_);
This way multiple findSafes can happen together.

Reply:
Read/write lock approach is efficient only in case there are significantly more reads than writes. In case the difference is not big, it is more expensive than a mutex.